dll/win32/browseui/regtreeoptions.cpp

```diff
@@ -1,59 +1,565 @@
+}
+
+void CRegTreeOptions::AddItemsFromRegistry(HKEY hKey, HTREEITEM hParent, HTREEITEM hInsertAfter)
+{
+    for (DWORD Index = 0, cchName;;)
+    {
+        WCHAR Name[MAX_PATH], Temp[MAX_PATH];
+        cchName = _countof(Name);
+        if (RegEnumKeyExW(hKey, Index++, Name, &cchName, NULL, NULL, NULL, NULL) != ERROR_SUCCESS)
+            break;
+
+        HKEY hItemKey, hCloseKey = NULL;
+        if (RegOpenKeyExW(hKey, Name, 0, KEY_READ, &hItemKey) != ERROR_SUCCESS)
+            continue;
+
+        TVINSERTSTRUCT tvis;
+        TVITEM &tvi = tvis.item;
+        tvi.mask = TVIF_PARAM | TVIF_TEXT | TVIF_IMAGE | TVIF_SELECTEDIMAGE;
+        tvi.lParam = (LPARAM)hItemKey;
+        UINT Type = TYPE_INVALID;
+        if (GetRegString(hItemKey, L"Type", Temp, _countof(Temp)) > 0)
+        {
+            if (!lstrcmpi(Temp, L"checkbox"))
+                Type = TYPE_CHECK;
+            if (!lstrcmpi(Temp, L"radio"))
+                Type = TYPE_RADIO;
+            if (!lstrcmpi(Temp, L"group"))
+                Type = TYPE_GROUP;
+        }
+        BOOL Valid = Type != TYPE_INVALID;
+
+        if (Type == TYPE_GROUP)
+        {
+            tvi.lParam = 0;
+            hCloseKey = hItemKey;
+            tvi.iImage = AddIconFromRegistry(m_hIL, hItemKey, L"Bitmap");
+        }
+        else if (Valid)
+        {
+            HRESULT hr = GetCheckState(hItemKey);
+            Valid = SUCCEEDED(hr);
+            tvi.iImage = (Type * STATEIMAGESPERTYPE) | (hr == S_OK);
+        }
+
+        if (Valid)
+        {
+            if (SUCCEEDED(SHLoadRegUIStringW(hItemKey, L"Text", Temp, _countof(Temp))))
+                tvi.pszText = Temp;
+            else
+                tvi.pszText = Name;
+#ifdef __REACTOS__ // Grayed is a ROS extension
+            DWORD grayed = GetRegDWORD(hItemKey, L"Grayed", DWORD(FALSE));
+            if (grayed)
+            {
+                tvi.lParam = (LPARAM)HKEY_GRAYED;
+                hCloseKey = hItemKey;
+                tvi.iImage += 2; // Use disabled state image
+                if (grayed & 4)
+                    tvi.iImage = (tvi.iImage & ~1) | ((grayed & 2) != 0); // Force a specific state
+            }
+#endif
+            tvi.iSelectedImage = tvi.iImage;
+            tvis.hParent = hParent;
+            tvis.hInsertAfter = hInsertAfter;
+            Valid = (tvi.hItem = TreeView_InsertItem(m_hTree, &tvis)) != NULL;
+        }
+
+        if (Valid && Type == TYPE_GROUP)
+        {
+            AddItemsFromRegistry(hItemKey, tvi.hItem, TVI_FIRST);
+            TreeView_Expand(m_hTree, tvi.hItem, TVE_EXPAND);
+        }
+        if (!Valid)
+            RegCloseKey(hItemKey);
+        if (hCloseKey && hItemKey != hCloseKey)
+            RegCloseKey(hCloseKey);
+    }
+    SendMessage(m_hTree, TVM_SORTCHILDREN, 0, (LPARAM)hParent);
+}
+
+HRESULT STDMETHODCALLTYPE CRegTreeOptions::InitTree(HWND hTV, HKEY hKey, LPCSTR SubKey, char const *param18)
+{
+    UNREFERENCED_PARAMETER(param18);
```

Comment:
Preferrably no arguments named like this, call it `pUnknown` or `pReserved` or something like it.

Reply:
That was from the existing code. I don't know what this parameter is for. I'll change it.